packages/next/src/server/app-render/create-error-handler.tsx

```diff
@@ -21,6 +21,27 @@ type SSRErrorHandler = (
 
 export type DigestedError = Error & { digest: string }
 
+/**
+ * Returns a digest for well-known Next.js errors, otherwise `undefined`. If a
+ * digest is returned this also means that the error does not need to be
+ * reported.
+ */
+export function getDigestForWellKnownError(error: unknown): string | undefined {
+  // If we're bailing out to CSR, we don't need to log the error.
+  if (isBailoutToCSRError(error)) return error.digest
+
+  // If this is a navigation error, we don't need to log the error.
+  if (isNextRouterError(error)) return error.digest
+
+  // If this error occurs, we know that we should be stopping the static
+  // render. This is only thrown in static generation when PPR is not enabled,
+  // which causes the whole page to be marked as dynamic. We don't need to
+  // tell the user about this error, as it's not actionable.
+  if (isDynamicServerError(error)) return error.digest
+
+  return undefined
+}
```

Comment:
I wonder if it's time to brand our errors so we don't have to do these delimited type checks. Is the semantic we're going for that we want to forward digests from errors that the framework created? This is important because we don't want a userland error with a digest property to hit this pathway. We can use instanceof checks I suppose. One thing to consider is that this function might only work in RSC layer b/c if any of these predicates check instance identity it won't work across the serialization boundary whereas a digest check will surivive since React forwards those.

Reply:
I guess it is also used on the client so they all have to check digest exclusively. This does mean it is relatively easy to create your own errors that appear to match these error types. Unlikely that this would actually happen but one way we can defend against that is for errors in RSC to never respect a passed in digest, only set the digest from some other source. Then the client can trust that a matching digest was always produced by the framework

Reply:
It's tricky, we currently also need to preserve digests from the `Cache` environment.